File: sdk/typescript/src/sandbox.ts
import type { Transport } from "./transport.js";
import type { Exec as ExecModel, FileInfo, FileList, Sandbox as SandboxModel, Snapshot as SnapshotModel } from "./types.js";
import { Snapshot } from "./snapshot.js";
import { collectExec, ExecStream, type ExecOptions, type ExecResult } from "./exec.js";
import { paginate } from "./pagination.js";

export class FileListing extends Array<FileInfo> {
  truncated = false;

  static get [Symbol.species](): ArrayConstructor {
    return Array;
  }
}

export class SandboxFiles {
  constructor(
    private readonly transport: Transport,
    private readonly ref: string,
  ) {}

  private path(suffix = ""): string {
    return `/v1/sandboxes/${this.transport.escape(this.ref)}/files${suffix}`;
  }

  private archive(): string {
    return `/v1/sandboxes/${this.transport.escape(this.ref)}/archive`;
  }

  async read(path: string): Promise<Uint8Array> {
    const response = await this.transport.request("GET", this.path(), { params: { path } });
    return new Uint8Array(await response.arrayBuffer());
  }

  async write(path: string, content: string | Uint8Array): Promise<void> {
    const raw = typeof content === "string" ? new TextEncoder().encode(content) : content;
    await this.transport.request("PUT", this.path(), { params: { path }, raw });
  }

  async stat(path: string): Promise<FileInfo> {
    const response = await this.transport.request("GET", this.path("/stat"), { params: { path } });
    return (await response.json()) as FileInfo;
  }

  async list(path: string): Promise<FileListing> {
    const response = await this.transport.request("GET", this.path("/list"), { params: { path } });
    const body = (await response.json()) as FileList;
    const listing = new FileListing();
    for (const item of body.items ?? []) listing.push(item);
    listing.truncated = Boolean(body.truncated);
    return listing;
  }

  async download(path: string): Promise<Uint8Array> {
    const response = await this.transport.request("GET", this.archive(), { params: { path } });
    return new Uint8Array(await response.arrayBuffer());
  }

  async upload(path: string, tar: Uint8Array): Promise<void> {
    await this.transport.request("PUT", this.archive(), { params: { path }, raw: tar });
  }
}

export class Sandbox {
  readonly files: SandboxFiles;
  forkFactory: (opts: Record<string, unknown>) => Promise<Sandbox> = () => {
    throw new Error("fork factory not wired");
  };

  constructor(
    private readonly transport: Transport,
    public raw: SandboxModel,
  ) {
    this.files = new SandboxFiles(transport, raw.id);
  }

  get id(): string {
    return this.raw.id;
  }

  get name(): string | undefined {
    return this.raw.name;
  }

  get state(): string {
    return this.raw.state;
  }

  get network(): string | undefined {
    return this.raw.network;
  }

  get ociRuntime(): string | undefined {
    return this.raw.oci_runtime;
  }

  private ref(): string {
    return this.transport.escape(this.raw.id);
  }

  private async replace(response: Response): Promise<this> {
    this.raw = (await response.json()) as SandboxModel;
    return this;
  }

  async refresh(): Promise<this> {
    return this.replace(await this.transport.request("GET", `/v1/sandboxes/${this.ref()}`));
  }

  async start(): Promise<this> {
    return this.replace(await this.transport.request("POST", `/v1/sandboxes/${this.ref()}/start`));
  }

  async stop(): Promise<this> {
    return this.replace(await this.transport.request("POST", `/v1/sandboxes/${this.ref()}/stop`));
  }

  async destroy(): Promise<this> {
    return this.replace(await this.transport.request("DELETE", `/v1/sandboxes/${this.ref()}`));
  }

  async snapshot(options: { name?: string } = {}): Promise<Snapshot> {
    const response = await this.transport.request("POST", `/v1/sandboxes/${this.ref()}/snapshots`, { body: options });
    return new Snapshot(this.transport, (await response.json()) as SnapshotModel, this.forkFactory);
  }

  async restore(snapshot: Snapshot | string): Promise<this> {
    const ref = typeof snapshot === "string" ? snapshot : snapshot.id;
    return this.replace(
      await this.transport.request("POST", `/v1/sandboxes/${this.ref()}/restore`, { body: { snapshot: ref } }),
    );
  }

  async exec(command: string | string[], options?: ExecOptions & { stream?: false }): Promise<ExecResult>;
  async exec(command: string | string[], options: ExecOptions & { stream: true }): Promise<ExecStream>;
  async exec(
    command: string | string[],
    options: ExecOptions & { stream?: boolean } = {},
  ): Promise<ExecResult | ExecStream> {
    const argv = typeof command === "string" ? ["sh", "-c", command] : [...command];
    const body: Record<string, unknown> = { command: argv };
    if (options.env) body.env = options.env;
    if (options.workingDir) body.working_dir = options.workingDir;
    if (options.user) body.user = options.user;
    const response = await this.transport.request("POST", `/v1/sandboxes/${this.ref()}/execs`, { body });
    const created = (await response.json()) as ExecModel;
    if (options.stream) return new ExecStream(this.transport, created.id);
    return collectExec(this.transport, created);
  }

  execs(filters: Record<string, string | number | undefined> = {}): AsyncGenerator<ExecModel> {
    return paginate(this.transport, `/v1/sandboxes/${this.ref()}/execs`, filters, (item: ExecModel) => item);
  }

  snapshots(filters: Record<string, string | number | undefined> = {}): AsyncGenerator<Snapshot> {
    return paginate(this.transport, `/v1/sandboxes/${this.ref()}/snapshots`, filters, (item: SnapshotModel) =>
      new Snapshot(this.transport, item, this.forkFactory),
    );
  }
}

File: sdk/python/src/orcal/sandbox.py
from . import models


class SandboxFiles:
    def __init__(self, client, ref):
        self._client = client
        self._ref = ref

    def _path(self, suffix=""):
        return f"/v1/sandboxes/{self._client._transport.escape(self._ref)}/files{suffix}"

    def read(self, path):
        return self._client._transport.request("GET", self._path(), params={"path": path}).content

    def write(self, path, content):
        payload = content.encode() if isinstance(content, str) else content
        self._client._transport.request("PUT", self._path(), params={"path": path}, content=payload)

    def stat(self, path):
        body = self._client._transport.request("GET", self._path("/stat"), params={"path": path}).json()
        return models.FileInfo(**body)

    def list(self, path):
        body = self._client._transport.request("GET", self._path("/list"), params={"path": path}).json()
        return [models.FileInfo(**entry) for entry in body.get("items", [])]

    def download(self, path):
        archive = f"/v1/sandboxes/{self._client._transport.escape(self._ref)}/archive"
        return self._client._transport.request("GET", archive, params={"path": path}).content

    def upload(self, path, tar_bytes):
        archive = f"/v1/sandboxes/{self._client._transport.escape(self._ref)}/archive"
        self._client._transport.request("PUT", archive, params={"path": path}, content=tar_bytes)


class Sandbox:
    def __init__(self, client, raw):
        self._client = client
        self.raw = raw
        self.files = SandboxFiles(client, raw.id)

    @classmethod
    def _from_payload(cls, client, payload):
        return cls(client, models.Sandbox(**payload))

    @property
    def id(self):
        return self.raw.id

    @property
    def name(self):
        return self.raw.name

    @property
    def state(self):
        return self.raw.state

    @property
    def network(self):
        return self.raw.network

    @property
    def oci_runtime(self):
        return getattr(self.raw, "oci_runtime", None)

    def _ref(self):
        return self._client._transport.escape(self.raw.id)

    def _replace(self, payload):
        self.raw = models.Sandbox(**payload)
        return self

    def refresh(self):
        return self._replace(self._client._transport.request("GET", f"/v1/sandboxes/{self._ref()}").json())

    def start(self):
        return self._replace(self._client._transport.request("POST", f"/v1/sandboxes/{self._ref()}/start").json())

    def stop(self):
        return self._replace(self._client._transport.request("POST", f"/v1/sandboxes/{self._ref()}/stop").json())

    def destroy(self):
        return self._replace(self._client._transport.request("DELETE", f"/v1/sandboxes/{self._ref()}").json())

    def snapshot(self, name=None):
        from .snapshot import Snapshot

        body = {} if name is None else {"name": name}
        payload = self._client._transport.request(
            "POST", f"/v1/sandboxes/{self._ref()}/snapshots", json=body
        ).json()
        return Snapshot._from_payload(self._client, payload)

    def restore(self, snapshot):
        ref = snapshot.id if hasattr(snapshot, "id") else snapshot
        payload = self._client._transport.request(
            "POST", f"/v1/sandboxes/{self._ref()}/restore", json={"snapshot": ref}
        ).json()
        return self._replace(payload)

    def exec(self, command, *, stream=False, env=None, working_dir=None, user=None):
        from .exec import ExecResult, ExecStream

        argv = ["sh", "-c", command] if isinstance(command, str) else list(command)
        body = {"command": argv}
        if env:
            body["env"] = env
        if working_dir:
            body["working_dir"] = working_dir
        if user:
            body["user"] = user
        created = self._client._transport.request("POST", f"/v1/sandboxes/{self._ref()}/execs", json=body).json()
        raw = models.Exec(**created)
        handle = ExecStream(self._client, raw.id)
        if stream:
            return handle
        out = []
        err = []
        for frame in handle:
            (out if frame.stream == "stdout" else err).append(frame.data)
        return ExecResult(
            id=handle.id,
            stdout=b"".join(out).decode("utf-8", errors="replace"),
            stderr=b"".join(err).decode("utf-8", errors="replace"),
            exit_code=handle.exit_code,
            truncated=handle.truncated,
            raw=raw,
            gaps=list(handle.gaps),
        )

    def execs(self, **filters):
        from ._pagination import paginate

        return paginate(
            self._client._transport,
            f"/v1/sandboxes/{self._ref()}/execs",
            filters,
            lambda item: models.Exec(**item),
        )

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc, tb):
        self.destroy()
        return False
